### ISSUE TITLE
python: KVSDir object should write to path it is initialized with

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#5308]
Problem: a KVSDir object acts like a dictionary for a  KVS directory when _reading_ from the KVS, but it discards its relative path and operates directly on the root when _writing_ to the KVS.

For example, this program uses a KVSDir to print a key from a job directory and then update the same key:
```python
#!/usr/bin/env python3

import sys
import flux.job
import flux.job.JobID as JobID

jobid = JobID(sys.argv[1])
key = sys.argv[2]
val = sys.argv[3]

handle = flux.Flux()
jobdir = flux.job.job_kvs(handle, jobid)
# This  refers to a key name in the job's KVS directory
print(jobdir[key])

# But this refers to a key name in the root directory
jobdir[key] = val
jobdir.commit()
```

```
$ flux ./foo.py  ƒ6acjkLQf jobspec xyz
{'resources': [{'type': 'slot', 'count': 1, 'with': [{'type': 'core', 'count': 1}], 'label': 'task'}], 'tasks': [{'command': ['hostname'], 'slot': 'task', 'count': {'per_slot': 1}}], 'attributes': {'system': {'duration': 0, 'cwd': '/nfshome/garlick', 'shell': {'options': {'rlimit': {'cpu': -1, 'fsize': -1, 'data': -1, 'stack': -1, 'core': 16384, 'nofile': 128000, 'as': -1, 'rss': -1, 'nproc': 8192}}}}}, 'version': 1}

# hmm no update
$ flux job info ƒ6acjkLQf jobspec
{"resources":[{"type":"slot","count":1,"with":[{"type":"core","count":1}],"label":"task"}],"tasks":[{"command":["hostname"],"slot":"task","count":{"per_slot":1}}],"attributes":{"system":{"duration":0,"cwd":"/nfshome/garlick","shell":{"options":{"rlimit":{"cpu":-1,"fsize":-1,"data":-1,"stack":-1,"core":16384,"nofile":128000,"as":-1,"rss":-1,"nproc":8192}}}}},"version":1}

# oops unexpected jobspec key in the root!
$ flux kvs get jobspec
"xyz"
```

**Comment by vsoch** [comment]
Thanks for opening this @garlick !

**Comment by garlick** [comment]
Thanks for finding it!

**Comment by chu11** [comment]
Looking through code history, this appears to be an oversight from the initial implementation in 2015.  Surprised it took so long to find :-)  Also no `path = ....` tests in `t/python/`, so should do that too.

**Comment by vsoch** [comment]
That's me, the bug finder!! :laughing: :bug: :lady_beetle: Actually more like I'm really good at breaking things. :hammer:

**Comment by vsoch** [comment]
Workaround - works! https://github.com/researchapps/test-kvs-commit/commit/9a8d1be520819b98350c8e89ba5c6076e3473a0c

Thanks for the help! This should work for our current needs-  it works for both the case that the key exists and the case that it does not.

**Comment by chu11** [comment]
note: looking through code, I think `fill()` and `mkdir()` on KVSDir probably also don't work.

**Comment by chu11** [comment]
thinking about this a bit ... this change has the potential to break a ton of code, if people are using it.  Which I don't know if people are.

i.e.

```
dir = KVSDir(handle, path="foo.bar")
dir["foo.bar.a"] = 1
dir.commit()
```

before this created "foo.bar.a" and if we change it'll create "foo.bar.foo.bar.a".

Perhaps we could ping a few code teams and see if they use KVSDir with a `path` argument and do writes?  If we can confirm a few teams don't use the path argument, then I'd feel better that this isn't going to break the world.

If a few teams are using it, then we perhaps should have a migration path to warn them of the change.

**Comment by vsoch** [comment]
If the bug hasn't been caught for 8 years, do you think people are using it?

**Comment by chu11** [comment]
> If the bug hasn't been caught for 8 years, do you think people are using it?

I agree low probability.  Two things made me pause.

1) With using `KVSDir` directly, I think this issue is a little less obvious.  You going through `job_kvs()` makes it a bit more obvious.  So it may have not been an obvious issue to any users using `KVSDir` directly.

2) (this is the bigger one I paused on) there's a `KVSDir` function called `fill()` which allows the user to pass in a Mapping and put each "key=val" mapping into the KVS.   It's not a dumb setter/getter that one would "just add" to a newly created class object.  I think there's a decent chance it was added by a user request OR added b/c of a specific need.
